modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -189,6 +101,58 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: Hasher] pr
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def joinTo(toJoin: P2PContext): EitherT[F, Throwable, Peer] =
+    twoWayHandshake(toJoin, none).attemptT
+      .flatMapF[Throwable, Peer] { p =>
+        clusterStorage
+          .getPeer(p.id)
+          .map(
+            Either.fromOption(
+              _,
+              FailedToJoin(p)
+            )
+          )
+      }
+
+  private def discoverFrom(peer: Peer): F[Set[Peer]] =
+    peerDiscovery
+      .discoverFrom(peer)
+      .handleErrorWith { e =>
+        MonadThrow[F].raiseError(FailedToDiscoverFrom(peer, e))
+      }
+
+  private def joinAndDiscover(toJoin: P2PContext): EitherT[F, Throwable, Set[P2PContext]] =
+    joinTo(toJoin)
+      .semiflatMap(discoverFrom)
+      .map(_.map(toP2PContext))
+
+  private def joinAll(peerToJoin: P2PContext): F[Unit] = {
+    type Agg = (Set[P2PContext], Set[P2PContext])
+    type Res = Unit
+
+    (Set(peerToJoin), Set.empty[P2PContext]).tailRecM {
+      case (toJoin, _) if toJoin.isEmpty => ().asRight[Agg].pure
+      case (toJoin, attempted) =>
+        for {
+          result <- toJoin.toList.parTraverse { peer =>
+            joinAndDiscover(peer).leftSemiflatMap { err =>
+              logger.warn(err)(s"Failed to join and discover from ${peer.show}")
+            }
+          }.value
+          discovered = result.toOption
+            .getOrElse(List.empty[Set[P2PContext]])
+          reduced = discovered.foldLeft(Set.empty[P2PContext]) {
+            case (acc, curr) =>
+              acc ++ curr
+          }
```

Comment:
1. for `discovered =`, you don't need to convert `result` to an option; `Either.getOrElse` does what you want
2. the above is because `result` is typed with `Unit` on the left of the `Either` -- seems a bit odd
3. for `reduced =`, that can be simplified to `discovered.foldLeft(Set.empty[P2PContext])(_ ++ _)`
 or you can simplify it further as `discovered.combineAll` if you `import cats.syntax.foldable._`

Reply:
Follow-up to the comment on the type of `result`, assuming you intend to keep the other functions returning an `EitherT` then maybe this is a bit cleaner:
```
discovered <- toJoin.toList.parTraverse { peer =>
  joinAndDiscover(peer).valueOrF(err =>
    logger.warn(err)(s"Failed to join and discover from ${peer.show}").as(Set.empty[P2PContext])
  )
}
```

Reply:
Changed to

```scala
          discovered <- toJoin.toList.parTraverse { peer =>
            joinAndDiscover(peer).valueOrF { err =>
              logger
                .warn(err)(s"Failed to join and discover from ${peer.show}")
                .as(Set.empty[P2PContext])
            }
          }
          reduced = discovered.combineAll
```